target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts

```diff
@@ -25,27 +25,6 @@
 	model = "OpenMesh A42";
 	compatible = "openmesh,a42", "qcom,ipq4019";
 
-	reserved-memory {
```

Comment:
Or maybe the u-boot adds it? Can you check if the /reserved-memory/ node gets added in /proc/device-tree/ even if the defintion is removed from the .dts?

I know that writing to 87e00000 - 88000000 can cause the system to crash.
Especially addresses starting from 87e80000 onwards. As this memory belongs to the trustzone hypervisor thats running in the background the whole time. 

(It's kinda sad that the all CPUs are booted in the SVC mode and not HVP. As we miss out on the virtualization extensions because of this)

Reply:
I've checked this previously on WPJ428 and it wasn't added. It should also not be added on A42 (but I cannot verify it right now from here). The region was removed from the WPJ428 because I couldn't show mkresin the bootup problem anymore which I had previously without it.

If this is really for the TZ then wouldn't it be better to have this reserved memory stored in a central place? Right now some IPQ40xx DTS add it and some not - and many of them add different regions.

Reply:
> I've checked this previously on WPJ428 and it wasn't added. It should also not be added on A42 (but I cannot verify it right now from here). The region was removed from the WPJ428 because I couldn't show mkresin the bootup problem anymore which I had previously without it.

I remember the grief tracing down the booting issue on the WPJ428 with @mehlis . I don't know if this is related or not. But back the WPJ428 would simply not boot (and not leave any messages on the serial console) and no matter what we tried it remained silent. Only after booting a modified rt-ac58u uboot the generated linux initramfs image from the LEDE-IPQ40XX would run. The reason why this didn't work was because the Compex u-boot would place the processed fdt into the reserved-memory area (since the initramfs image was slightly too big). And this prevented the kernel from parsing the fdt (apart from the early_fdt). Once the reserved-memory node (rsvd1@87000000) was removed the device booted the images... Since @mehlis didn't want to remove the reserve-memory node he added the "set fdt_high 83000000" u-boot env variable, so this was no longer a problem.


About the reserved-memory node. I managed to get the information from @yeryomin on the mailing-list. He was kind enough to make a patch with the proper memory node names.
This was added to:

https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/864-07-dts-ipq4019-ap-dk01.1-c1-add-spi-and-ram-nodes.patch

If you want you could run this experiment in the u-boot on the WPJ428 and see if you get the same result (I c&p this from my RT-AC58U with the stock u-boot) :
    (IPQ40xx) # mw 87e70000 0

    (IPQ40xx) # md 87e80000
    87e80000: e59ff364 e59ff364 e59ff364 e59ff364    d...d...d...d...
    87e80010: e59ff364 e59ff364 e59ff360 e59ff35c    d...d...`...\...
    87e80020: ee0d0f50 ee0def90 e59f0354 e12fff30    P.......T...0./.
    87e80030: e3500000 1a0000ba ee1d0f50 e59f3344    ..P.....P...D3..
    87e80040: e5933000 e3530000 0a000086 eaffffff    .0....S.........
    87e80050: e24f0008 e59f1330 e1500001 1afffffe    ..O.0.....P.....
    87e80060: e59f1328 e5911000 ee100fb0 e2000cff    (...............
    87e80070: e1a00440 e0000190 ee101fb0 e20110ff    @...............
    87e80080: e0810000 e59f1308 e3a02001 e7812100    ......... ...!..
    87e80090: e59f1300 ee102fb0 e20220ff e2822001    ...../... ... ..
    87e800a0: e1910f9f e3500000 01810f92 03500000    ......P.......P.
    87e800b0: 1afffffa f57ff05f ee100fb0 e2000cff    ...._...........
    87e800c0: e1a00440 e59f12d0 e7912100 e2422001    @........!... B.
    87e800d0: e7812100 e59f12b4 e5911000 ee100fb0    .!..............
    87e800e0: e2000cff e1a00440 e0000190 ee101fb0    ....@...........
    87e800f0: e20110ff e0810000 e59f12a4 e7911100    ................
    (IPQ40xx) # mw 87e80000 0
    data abort  
    pc : [<873149f8>]          lr : [<87308578>]
    sp : 86edfc28  ip : 86ef4412     fp : 00000000
    r10: 00000000  r9 : 00000000     r8 : 86edff68
    r7 : 00000003  r6 : 8737e624     r5 : 86ef4420  r4 : 8736c154
    r3 : 00000000  r2 : 00000010     r1 : 00000000  r0 : 00000000
    Flags: nZCv  IRQs off  FIQs off  Mode SVC_32
    Resetting CPU ...

    resetting ...

> If this is really for the TZ then wouldn't it be better to have this reserved memory stored in a central place? Right now some IPQ40xx DTS add it and some not - and many of them add different regions.

Some u-boots do read/add or fiddle with the reserved-memory / memory nodes.
https://github.com/riptidewave93/meraki-uboot/blob/mr33-20170427/board/qcom/ipq40xx_cdp/ipq40xx_cdp.c#L764
(depends on whenever CONFIG_QCA_APPSBL_DLOAD is set or not).

so the question becomes: can the reserved-memory nodes be stored in centrally?
or will this lead to devices needing /delete-node/ rsvd0; hacks?

(Also changing the qcom-ipq4019.dtsi is difficult since the file is upstream, so any patches to the file should be sent to linux-arm-msm@vger.kernel.org as well)

Reply:
Theoretically, different boards may have different reserved memory map set.
That's why I didn't add that full set into common ipq40xx dtsi.
Practically, only the time will show...

Reply:
Both WPJ428 and A42 crashed with the mw and both didn't create the reserved memory entries in the device tree. Now comes the biggest question - how to find out what memory regions are reserved for what.

And to the u-boot code. The meraki code also doesn't create the node but just adjusts it. I have also checked the A42 and it would use the same code when CONFIG_QCA_APPSBL_DLOAD would have been enabled. It transforms (in theory):

           reserved-memory {
                   #address-cells = <0x1>;
                   #size-cells = <0x1>;
                   ranges;
    
                  /* apps_bl, sbl? */
                   rsvd1@87000000 {
                           reg = <0x87000000 0x500000>;
                           no-map;
                   };
    
                   /* cnss_debug? */
                   wifi_dump@87500000 {
                           reg = <0x87500000 0x600000>;
                           no-map;
                   };
    
                   /* cpu_context_dump, tz_apps, smem, tz? */
                   rsvd2@87B00000 {
                           reg = <0x87b00000 0x500000>;
                           no-map;
                   };
           };

To

           reserved-memory {
                   #address-cells = <0x1>;
                   #size-cells = <0x1>;
                   ranges;
    
                   /* cpu_context_dump, tz_apps, smem, tz? */
                   rsvd1@87b00000 {
                           reg = <0x87b00000 0x500000>;
                           no-map;
                   };
           };

I find it also quite interesting that all these nodes + reset_crashdump are necessary to reach reset_crashdump. But since none of the code is actually active, it doesn't seem to be important at all.

I will remove this commit now but this is definitely something where we are missing some information to make a decision about how to handle it.

Reply:
@yeryomin could you please check https://github.com/openwrt/openwrt/pull/810/commits/7dcbd3180d6f74427b0a3d1ff3c446c6f390c1bd ("ipq40xx: Add reserved memory for WPJ428") makes sense? You seem to know a little bit more about it then me.

Reply:
> I find it also quite interesting that all these nodes + reset_crashdump are necessary to reach reset_crashdump. But since none of the code is actually active, it doesn't seem to be important at all.

>I will remove this commit now but this is definitely something where we are missing some information to make a decision about how to handle it.

Just a (related) note about tz: The tz can be accessed through the qcom-scm.c
https://github.com/torvalds/linux/blob/master/drivers/firmware/qcom_scm.c.
Nowadays, it would only require adding the scm firmware node in the dts.

I used to have this enabled:
https://github.com/chunkeey/LEDE-IPQ40XX/commit/4c75910103f2fbe934f299d0bade25226feab281#diff-75a9ee647eceaa46bea8e9697abdfee2

And this firmware node was also part of the https://patchwork.kernel.org/patch/10189259/
that became
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/101-ARM-dts-ipq4019-Add-a-few-peripheral-nodes.patch
However, the node has been removed from the patch by I think @mkresin (but it could also have been  john?). so there's probably no point in adding it back.

Reply:
@ecsv it makes sense for sure